Adding documentation about curvilinear coordinates (#181)
gyselax/gyselalibxx: docs/mathematical_and_physical_conventions.md

# Mathematical and physical conventions

The conventions adopted in Gyselalib++ for describing physical quantities such as scalar fields, vector fields and tensors are detailed hereafter. 

## Contents

-  [On the use of curvilinear coordinates](#docs_mathematical_and_physical_conventions__On_the_use_of_curvilinear_coordinates)
-  [Metric tensor](#docs_mathematical_and_physical_conventions__Metric_tensor)
-  [Jacobian](#docs_mathematical_and_physical_conventions__Jacobian)
-  [Differential operators](#docs_mathematical_and_physical_conventions__Differential_operators)

## On the use of curvilinear coordinates

Let us consider a system of coordinates denoted by $`\{q^i\}_{ i \in [1, N]}`$, where $`i`$ is an integer quantity and $`N`$ is the dimension of the space mapped by the $`\{q^i\}`$ coordinates. For the sake of simplicity we use in the following the compact $`\{q^i\}`$ notation to denote the set $`\{q^i\}_{ i \in [1, N]}`$. The $`\{q^i\}`$ coordinate system can be derived from Cartesian coordinates---that we denote by $`\{x^i\}`$--- from the invertible transformation 

```math
q^i = q^i(x^1, \ldots , x^N) \quad \forall i \in [1, N].
```

> Note that even though an inverse of the functions $`q^i(x^1, \ldots , x^N)`$ must exist so that the coordinate transform is valid, it is not always possible to write an explicit formula for such inverse. The surface defined by $`q^i = \text{constant}`$ is called a *coordinate surface*, the intersection of these surfaces define *coordinate curves*. A *coordinate axis* is locally defined as the axis tangent to a given coordinate curve. As their name suggests, coordinate curves do not form straight lines in general. An illustration for these geometrical concepts is given in the following figure.

Reviewer: For me, the definition you give of "coordinate curve" is not so clear. I propose:
One "i coordinate surface" is defined as as $`q^i = \text{constant}`$. One "i coordinate curve" is defined as the intersection of a serie of j coordinate surfaces $\forall j \in [1, N]$ with $ j\neq i$. A "i coordinate axis" is defined as the normal of the "i coordinate surface".

Author: I agree with you except for the definition of the coordinate axis: this axis is defined as the tangent of the coordinate curve, not as the normal of the coordinate surface (see https://en.wikipedia.org/wiki/Curvilinear_coordinates, or simply below)
![image](https://github.com/user-attachments/assets/97c9a14f-fe1f-4113-a294-7125c915d39d)

The two (tangents and normal axis) are not the same in general, in fact it this difference that makes the covariant and contravariant bases different (cf. the same wiki page).

I've opened https://github.com/gyselax/gyselalibxx/pull/219 to update the doc according to your suggestions

![Coordinate curves, surfaces and axes in a three-dimensional space](./curvilinear_coordinates_def.png "")

The position of any point in space can be written as

```math

> \overrightarrow{r} = x^i \mathbf{e}_i,

Reviewer: Why not simply \overrightarrow{x}?

Author: https://github.com/gyselax/gyselalibxx/pull/219

![Geometrical interpretation of the contravariant and covariant bases vectors](./curvilinear_coordinates_contravariant_covariant_bases.png)

Note that neither contravariant nor covariant bases form orthonormal vector sets in general. Additionally note that in the case of Cartesian coordinates covariant and contravariant bases are the same. In general, the following property holds 
```math
\mathbf{b}^i \cdot \mathbf{b}_j = \delta_{ij}, 
```

with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`\mathbf{A} \in \mathbb{R}^N`$. This vector may be expressed in either covariant or contravariant bases as

```math
\mathbf{A} = A_i \mathbf{b}^i = A^i \mathbf{b}_i, 
```

Where 

> - $`\{A_i\}`$ are the components of $`\mathbf{A}`$ in *the covariant basis*. We call these *the covariant components* of $`A`$;

Reviewer 2: ```suggestion
- $`\{A_i\}`$ are the components of $`\mathbf{A}`$ in *the covariant basis*. We call these *the covariant components* of $`\mathbf{A}`$;
```

> - $`\{A^i\}`$ are the components of $`\mathbf{A}`$ in *the contravariant basis*. We call these *the contravariant components* of $`A`$;

Reviewer 2: ```suggestion
- $`\{A^i\}`$ are the components of $`\mathbf{A}`$ in *the contravariant basis*. We call these *the contravariant components* of $`\mathbf{A}`$;
```

## Metric tensor

The *metric tensor* $`G`$ associated with the set of coordinates $`\{q^i\}`$ is defined from its components $`g_{ij}`$ as

```math
g_{ij} = \mathbf{b}_i \cdot \mathbf{b}_j,
```

and the *inverse metric tensor* $`G^{-1}`$, whose components are written as $`g^{ij}`$, is defined as

```math
g^{ij} = \mathbf{b}^i \cdot \mathbf{b}^j.
```

A property of the metric tensor is that it can be used to transform contravariant components of a vector into covariant components and conversely, as 

```math
\begin{aligned}
A_i =& g_{ij}A^j, \\
A^i =& g^{ij}A_j.
\end{aligned}
```

An easy way to get the above formulas right is to remember that the metric tensor can be used to "raise" or "lower" indices. 

## Jacobian

### Curvilinear to Cartesian

> The *Jacobian* $`J`$ associated with the transformation from the curvilinear coordinates $`\{q^i\}`$ to the Cartesian coordinates $`\{x^i\}`$ is defined from its components $`J\left.^{i}_{\;j}\right.`$ as 

Reviewer 2: ```suggestion
The *Jacobian* matrix $`J`$ associated with the transformation from the curvilinear coordinates $`\{q^i\}`$ to the Cartesian coordinates $`\{x^i\}`$ is defined from its components $`J\left.^{i}_{\;j}\right.`$ as 
```
I am quibbling. Even I doubt they will mix with the Jacobian determinant.

```math
J\left.^{i}_{\;j}\right. = \frac{\partial x^i}{\partial q^j},
```

while the inverse Jacobian $`J^{-1}`$ is defined by its components $`J^{-1}\left.^{i}_{\;j}\right.`$ as 

```math
J^{-1}\left.^{i}_{\;j}\right. = \frac{\partial q^i}{\partial x^j}.
```

Both Jacobian and metric tensor relate to each other as

```math
G = J^{T}J.
```

> The Jacobian of a curvilinear coordinate transformation can be used to relate the components of a vector expressed in the Cartesian basis $`\{\mathbf{e}_i\}`$ to the components of the vector expressed in the contravariant basis $`\{\mathbf{b}_i\}`$ associated with the curvilinear coordinate system. More precisely, let us write a vector $`A`$ as 

Reviewer 2: ```suggestion
The Jacobian of a curvilinear coordinate transformation can be used to relate the components of a vector expressed in the Cartesian basis $`\{\mathbf{e}_i\}`$ to the components of the vector expressed in the contravariant basis $`\{\mathbf{b}_i\}`$ associated with the curvilinear coordinate system. More precisely, let us write a vector $`\mathbf{A}`$ as 
```

```math

> A = A^i_\text{c} \mathbf{e}_i = A^i \mathbf{b}_i, 

Reviewer 2: ```suggestion
\mathbf{A} = A^i_\text{c} \mathbf{e}_i = A^i \mathbf{b}_i, 
```

```

Where the "c" subscript indicates that the considered components is computed in the Cartesian basis. Note that we used the contravariant basis in the expression above. It can be shown using the chain rule that the following equality holds

```math
A^i_\text{c} = J\left.^{i}_{\;j}\right. A^j. 
```

> Note that relating the components of the vector $`A`$ expressed in the covariant basis is less straightforward as this involves the metric tensor, i.e.  

Reviewer 2: ```suggestion
Note that relating the components of the vector $`\mathbf{A}`$ expressed in the covariant basis is less straightforward as this involves the metric tensor, i.e.  
```

```math
A^i_\text{c} = J\left.^{i}_{\;j}\right. g^{jk}A_k. 
```

### Curvilinear to curvilinear

Let us now consider the more general case where one seeks to relate two curvilinear coordinate systems $`\{q^i\}`$ and $`\{p^i\}`$. We denote by $`J\{q\to p\}`$ the Jacobian of the transformation from the coordinate system $`\{q^i\}`$ to $`\{p^i\}`$. The components of the Jacobian and of its inverse are defined similarly as above as 

```math
J\left.^{i}_{\;j}\right.\{q\to p\} = \frac{\partial p^i}{\partial q^j}, \quad J^{-1}\left.^{i}_{\;j}\right.\{p\to q\} = \frac{\partial q^i}{\partial p^j}. 
```

> Here one may note that $`J^{-1}\left.^{i}_{\;j}\right.\{p\to q\} = J\left.^{i}_{\;j}\right.\{q\to p\}`$. Let us now write $`\{\mathbf{b}_i\}`$ (resp. $`\{\mathbf{b}^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{q^i\}`$, and $`\{\mathbf{c}_i\}`$ (resp. $`\{\mathbf{c}^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{p^i\}`$. Let us now express a vector $`A`$ in these bases as 

Reviewer 2: ```suggestion
Here one may note that $`J^{-1}\left.^{i}_{\;j}\right.\{p\to q\} = J\left.^{i}_{\;j}\right.\{q\to p\}`$. Let us now write $`\{\mathbf{b}_i\}`$ (resp. $`\{\mathbf{b}^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{q^i\}`$, and $`\{\mathbf{c}_i\}`$ (resp. $`\{\mathbf{c}^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{p^i\}`$. Let us now express a vector $`\mathbf{A}`$ in these bases as 
```

```math

> A = A^i\{p\} \mathbf{b}_i = A_i\{p\} \mathbf{b}^i = A^i\{q\} \mathbf{c}_i = A_i\{q\} \mathbf{c}^i, 

Reviewer 2: ```suggestion
\mathbf{A} = A^i\{p\} \mathbf{b}_i = A_i\{p\} \mathbf{b}^i = A^i\{q\} \mathbf{c}_i = A_i\{q\} \mathbf{c}^i, 
```

```

> where $`A^i\{p\}`$ (resp. $`A_i\{p\}`$) refers to the $`i`$-th contravariant (resp. covariant) component of $`A`$ expressed in the vector basis associated with coordinates $`\{p^i\}`$, and similarly $`A^i\{q\}`$ (resp. $`A_i\{q\}`$) refers to the $`i`$-th contravariant (resp. covariant) component of $`A`$ expressed in the vector basis associated with coordinates $`\{q^i\}`$. We have

Reviewer 2: ```suggestion
where $`A^i\{p\}`$ (resp. $`A_i\{p\}`$) refers to the $`i`$-th contravariant (resp. covariant) component of $`\mathbf{A}`$ expressed in the vector basis associated with coordinates $`\{p^i\}`$, and similarly $`A^i\{q\}`$ (resp. $`A_i\{q\}`$) refers to the $`i`$-th contravariant (resp. covariant) component of $`\mathbf{A}`$ expressed in the vector basis associated with coordinates $`\{q^i\}`$. We have
```

```math
A^i\{p\} = J\left.^{i}_{\;j}\right.\{q\to p\} A^j\{q\}. 
```

> By introducing the metric tensors $`G\{p\}`$  and $`G\{q\}`$ associated with both curvilinear coordinates system, and by writing $`g_{ij}\{p\}`$ and $`g_{ij}\{q\}`$ their elements ($`g^{ij}\{p\}`$ and $`g^{ij}\{q\}`$ the elements of their inverse) one relates the components of $`A`$ expressed in covariant vector basis associated with both curvilinear coordinates system as 

Reviewer 2: ```suggestion
By introducing the metric tensors $`G\{p\}`$  and $`G\{q\}`$ associated with both curvilinear coordinates system, and by writing $`g_{ij}\{p\}`$ and $`g_{ij}\{q\}`$ their elements ($`g^{ij}\{p\}`$ and $`g^{ij}\{q\}`$ the elements of their inverse) one relates the components of $`\mathbf{A}`$ expressed in covariant vector basis associated with both curvilinear coordinates system as 
```

# Mathematical and physical conventions

The conventions adopted in gyselalibxx for describing physical quantities such as scalar fields, vector fields and tensors are detailed hereafter. 

## Contents

-  [On the use of curvilinear coordinates](#docs_mathematical_and_physical_conventions__On_the_use_of_curvilinear_coordinates)
-  [Metric tensor](#docs_mathematical_and_physical_conventions__Metric_tensor)
-  [Jacobian](#docs_mathematical_and_physical_conventions__Jacobian)
-  [Differential operators](#docs_mathematical_and_physical_conventions__Differential_operators)

## On the use of curvilinear coordinates

> Let us consider a system of coordinates denoted by $`\{q^i\}_{ i \in [1, N]}`$, where $`i`$ is an integer quantity and $`N`$ is the dimension of the space mapped by the $`\{q^i\}`$ coordinates. For the sake of simplicity we use in the following the compact $`\{q^i\}`$ notation to denote the set $`\{q^i\}_{ i \in [1, N]}`$. The $`\{q^i\}`$ coordinate system can be derived from Cartesian coordinates---that we denote by $`\{x^i\}`$--- from the invertible transformation 

Reviewer 3: > from the invertible transformation 

Maybe we should specify that although the transformation is invertible in the mathematical sense it is not necessarily analytically invertible?

Author: done 6b961e3aa139c48fd9d3c956e643c0ace9ba67eb

```math
q^i = q^i(x^1, \ldots , x^N) \quad \forall i \in [1, N].
```

> The surface defined by $`q^i = \text{constant}`$ is called a *coordinate surface*, the intersection of these surfaces define *coordinates curves*. A *coordinate axis* is locally defined as the axis tangent to a given coordinate curve. **TODO ADD IMAGE**. As their name suggest, coordinates curves do not form straight lines in general. The position of any point in space can be written as

Reviewer 3: ```suggestion
The surface defined by $`q^i = \text{constant}`$ is called a *coordinate surface*, the intersection of these surfaces define *coordinate curves*. A *coordinate axis* is locally defined as the axis tangent to a given coordinate curve. **TODO ADD IMAGE**. As their name suggests, coordinate curves do not form straight lines in general. The position of any point in space can be written as
```

```math

> \mathbf{r} = x^i e_i,

Reviewer 3: I think we should avoid the letter `r` avoid possible misreading

Reviewer 2: I agree, why not "q" or "x" ? or "p" for point?

Author: I would suggest to keep the letter "r", but to add an arrow above it to indicate that it is the vector position we consider. This notation is very common in physics textbooks, besides "q", "x" or "p" are already used somewhere else in the note.

Author: see a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

```

> where we introduced the unit vectors of the orthonormal Cartesian basis $`\{e_i\}`$. We rely on the Einstein summation in the above equation and in the following, i.e. we consider that repeated indices indicate a summation.

Reviewer 2: I do not know if it is interesting to add a Wikipedia link on "Einstein summation" (https://en.wikipedia.org/wiki/Einstein_notation)

Author: yes definitely, done in d8074e2b385c418a3845e0b21507acb12b172937

> where we introduced the unit vectors of the orthonormal Cartesian basis $`\{e_i\}`$. We rely on the Einstein summation in the above equation and in the following, i.e. we consider that repeated indices indicate a summation.

Reviewer 2: $`\{e_i\}`$ are covariant or contravariant? 
Does the definition makes sense here?

Author: Here ${e_i}$ is a Cartesian basis, there is no distinction between covariant and contravariant bases, meaning $e_i = e^i$

Author: I' going to add this as a remark

Author: done 4d2d015040128e3a6d99116e6100de1f69b6f894

## Contravariant and covariant basis
Let us define the *contravariant basis* $`\{b_i\}`$ associated to the $`\{q^i\}`$ coordinates by

```math
b_i = \frac{\partial \mathbf{r}}{\partial q^i} = \frac{\partial x^j}{\partial q^i} e_j,
```

and the *covariant basis* $`\{b^i\}`$ by

```math

> b^i = \nabla q^i = \frac{\partial q^i}{\partial x^j} e_j.

Reviewer 2: The covariant and the contravariant bases are defined on the covariant vectors?

Reviewer 3: Where did you see the term "covariant vectors" in the doc?

```

> These two basis are *basis*, in the sense that the $`b_i`$ and $`b^i`$ vectors depend on the considered position in space, i.e. 

Reviewer 3: ```suggestion
These two bases are *bases*, in the sense that the $`b_i`$ and $`b^i`$ vectors depend on the considered position in space, i.e. 
```

Author: done a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

> These two basis are *basis*, in the sense that the $`b_i`$ and $`b^i`$ vectors depend on the considered position in space, i.e. 

Reviewer 3: I am not sure about your definition of a basis as "a vector that depends on the position in space"

Author: sorry typo, fixed in a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

we have $`b_i(x^1, \ldots , x^N)`$ and $`b^i(x^1, \ldots , x^N)`$. A geometrical interpretation of these two basis reads as follows. The *covariant* unit vector $`b^i`$ is orthogonal to the coordinate surface $`q^i = \text{constant}`$, while the *contravariant* unit vector $`b_i`$ is locally tangent to the coordinate curve associated with the $`q^i`$ coordinate **TODO ADD IMAGE**. Note that neither contravariant nor covariant basis form orthonormal vector sets in general, but the following property holds 
```math
b^i \cdot b_j = \delta_{ij}, 
```

with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`A`$. This vector may be expressed in either contravariant or covariant basis as

```math
A = A_i b^i = A^i b_i, 
```

Where 
- $`\{A_i\}`$ are the components of $`A`$ in the covariant basis. We call these *the contravariant components* of $`A`$;

> - $`\{A^i\}`$ are the components of $`A`$ in the contravariant basis. We call these *the covariant components* of $`A`$;

Reviewer 3: Strange to use italics for the components if the keyword in the code refers to the basis

Author: I italicized here to emphasize that these sentence define the contravariant and covariant components of a vector, not to indicate what the convention is in the code concerning the get<R_cov> function (this info is to be added in another pr)

- Covariant basis vectors scale "with" the coordinates, meaning the new covariant basis $`\{\tilde b^i\}`$ verifies $`\tilde b^i = \alpha b^i`$ (hence the name *covariant basis*);
- Contravariant basis vectors scale "in opposition with" the coordinates, meaning $`\tilde b_i = \alpha^{-1} b_i`$ (hence the name *contravariant basis*).

Given that $`A`$ must remain unchanged under such transformation of the coordinate system, its components in the covariant or contravariant vector basis must scale inversely to the considered basis, i.e. a contravariant component $`A_i`$ must be multiplied with a covariant vector basis $`b^i`$ in the expression of $`A`$.

## Metric tensor

The *metric tensor* $`G`$ associated with the set of coordinates $`\{q^i\}`$ is defined from its components $`g_{ij}`$ as

```math
g_{ij} = b_i \cdot b_j,
```

and the *inverse metric tensor* $`G^{-1}`$, whose components are written as $`g^{ij}`$, is defined as

```math
g^{ij} = b^i \cdot b^j.
```

A property of the metric tensor is that it can be used to transform contravariant components of a vector into covariant components and conversely, as 

```math

> A_i = g_{ij}A^j, \\
> A^i = g^{ij}A_j. \\

Reviewer 3: Something extra is needed for a multi-line equation
```suggestion
\begin{aligned}
A_i =& g_{ij}A^j, \\
A^i =& g^{ij}A_j.
\end{aligned}
```

Author: thanks ! a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

```

An easy way to get the above formulas right is to remember that the metric tensor can be used to "raise" or "lower" indices. 

## Jacobian

### Curvilinear to Cartesian

The *Jacobian* $`J`$ associated with the transformation from the curvilinear coordinates $`\{q^i\}`$ to the Cartesian coordinates $`\{x^i\}`$ is defined from its components $`J_{ij}`$ as 

```math

> J_{ij} = \frac{\partial x^i}{\partial q^j},

Reviewer 3: I would write:
$J^i_j$ with $i$ a covariant component to match $x^i$ and $j$ a contravariant component to match $\partial_{q^j}$

Author: done a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

```

while the inverse Jacobian $`J^{-1}`$ is defined by its components $`J^{-1}_{ij}`$ as 

```math

> J^{-1}_{ij} = \frac{\partial q^i}{\partial x^j}.

Reviewer 3: ```suggestion
J^{-1}^i\._{j} = \frac{\partial q^i}{\partial x^j}.
```

Author: done a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

```

Both Jacobian and metric tensor relate to each other as

```math
G = J^{T}J.
```

> The Jacobian of a curvilinear coordinate transform can be used to relate the components of a vector expressed in the Cartesian basis $`\{e_i\}`$ to the components of the vector expressed in the contravariant basis $`\{b_i\}`$ associated with the curvilinear coordinate system. More precisely, let us write a vector $`A`$ as 

Reviewer 3: ```suggestion
The Jacobian of a curvilinear coordinate transformation can be used to relate the components of a vector expressed in the Cartesian basis $`\{e_i\}`$ to the components of the vector expressed in the contravariant basis $`\{b_i\}`$ associated with the curvilinear coordinate system. More precisely, let us write a vector $`A`$ as 
```

```math
A = A^i_\text{c} e_i = A^i b_i, 
```

Where the "c" subscript indicates that the considered components is computed in the Cartesian basis. It can be shown using the chain rule that the following equality holds

```math

> A^i_\text{c} = J_{ij} A^j. 

Reviewer 3: ```suggestion
A^i_\text{c} = J^i_{j} A^j. 
```

Author: done a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

```

Note that relating the components of the vector $`A`$ expressed in the covariant basis is less straightforward as this involves the metric tensor, i.e.  

```math
A^i_\text{c} = J_{ij} g^{jk}A_k. 

> ```

Author: This could be adressed in another PR, see https://github.com/gyselax/gyselalibxx/issues/182

### Curvilinear to curvilinear

Let us now consider the more general case where one seeks to relate two curvilinear coordinate system $`\{q^i\}`$ and $`\{p^i\}`$. We denote by $`J\{q\to p\}`$ the Jacobian of the transformation from the coordinate system $`\{q^i\}`$ to $`\{p^i\}`$. The components of the Jacobian and of its inverse are defined similarly as above as 

```math

> J_{ij}\{q\to p\} = \frac{\partial p^i}{\partial q^j}, \quad J_{ij}^{-1}\{p\to q\} = \frac{\partial q^i}{\partial p^j}. 

Reviewer 3: ```suggestion
J^i_{j}\{q\to p\} = \frac{\partial p^i}{\partial q^j}, \quad (J^{-1})^i_{j}\{p\to q\} = \frac{\partial q^i}{\partial p^j}. 
```

Author: done a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

```

> Here one may note that $`J_{ij}^{-1}\{p\to q\} = J_{ij}\{q\to p\}`$. Let us now write $`\{b_i\}`$ (resp. $`\{b^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{q^i\}`$, and $`\{c_i\}`$ (resp. $`\{c^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{p^i\}`$. Let us now express a vector $`A`$ in these basis as 

Reviewer 3: ```suggestion
Here one may note that $`J_{ij}^{-1}\{p\to q\} = J_{ij}\{q\to p\}`$. Let us now write $`\{b_i\}`$ (resp. $`\{b^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{q^i\}`$, and $`\{c_i\}`$ (resp. $`\{c^i\}`$) the contravariant (resp. covariant) vector basis associated with coordinates $`\{p^i\}`$. Let us now express a vector $`A`$ in these bases as 
```

Author: done a9b63905d81bfb51ee7a7cf3d26939df5a4abffc

```math
A = A^i\{p\} b_i = A_i\{p\} b^i = A^i\{q\} c_i = A_i\{q\} c^i, 
```

where $`A^i\{p\}`$ (resp. $`A_i\{p\}`$) refers to the $`i`$-th covariant (resp. contravariant) component of $`A`$ expressed in the vector basis associated with coordinates $`\{p^i\}`$, and similarly $`A^i\{q\}`$ (resp. $`A_i\{q\}`$) refers to the $`i`$-th covariant (resp. contravariant) component of $`A`$ expressed in the vector basis associated with coordinates $`\{q^i\}`$. We have

```math

> A^i\{p\} = J_{ij}\{q\to p\} A^j\{p\}. 

Reviewer 3: Are you sure here? Is it not
```suggestion
A^i\{p\} = J_{ij}\{q\to p\} A^j\{q\}. 
```

```

By introducing the metric tensors $`G\{p\}`$  and $`G\{q\}`$ associated with both curvilinear coordinates system, and by writing $`g_{ij}\{p\}`$ and $`g_{ij}\{q\}`$ their elements ($`g^{ij}\{p\}`$ and $`g^{ij}\{q\}`$ the elements of their inverse) one relates the components of $`A`$ expressed in covariant vector basis associated with both curvilinear coordinates system as 

```math

> g^{ik}\{p\}A_k\{p\} = J_{ij}\{q\to p\}g^{jl}\{p\} A_l\{p\}. 

Reviewer 3: Shouldn't it be?
```suggestion
g^{ik}\{p\}A_k\{p\} = J_{ij}\{q\to p\}g^{jl}\{q\} A_l\{q\}. 
```

```

## Differential operators

Hereafter are expressed differential operators in a curvilinear coordinate system $`\{q^i\}`$. 

### Gradient

Let us consider a scalar field $`f`$. The gradient $`\nabla f`$ of such field is defined in the Cartesian basis $`\{e_i\}`$ as

```math
\nabla f = \frac{\partial f}{\partial x^i} e_i. 
```

This quantity can be expressed in the both contravariant $`\{b_i\}`$ and covariant $`\{b^i\}`$ vector basis as

```math
\nabla f = \frac{\partial f}{\partial q^i} b^i = g^{ij} \frac{\partial f}{\partial q^j} b_i. 

> ```

Reviewer 2: ![image](https://github.com/user-attachments/assets/4db9ec32-85e8-4038-8257-84a72bc7dff5)

I am confused :sweat_smile: 
So the _covariant basis_ is on the _contravariant vectors_?

Reviewer 2: Can you precise it somewhere for the people like me? 
Or do a table
 
vector               cov           contrav
basis                 contrav   cov
components   cov           contrav

? :sweat_smile:

Author: what do you mean the the covariant basis is on the contravariant vectors?

Reviewer 2: I dont have a lot of materials to refer. I am mainly using this wikipedia page where they talk about covariant vectors and contravariant vectors. The vectors in the covariants basis corresponds to the contravariants vectors in this wikipedia page?

Reviewer 3: > I am mainly using this wikipedia page
Do you have a link to share?

I found this wiki page:
https://en.wikipedia.org/wiki/Curvilinear_coordinates#Covariant_and_contravariant_bases
which uses similar terms to what we are using here. And this wiki page:
https://en.wikipedia.org/wiki/Covariance_and_contravariance_of_vectors
which despite being called "Covariance and contravariance of vectors" does not define a "covariant vector" or a "contravariant vector". It does mention "vectors" and "covectors" which are defined on the co/contravariant bases.

But I don't think either of these pages are the one that you have screenshotted above.

Author: I fixed a naming convetion error in the note, now everything should be consistent with what can be found on wikipedia af2c1fd60b0cb6b6f306467feca607aade9b6547

> 

Reviewer 3: Maybe good to mention that it is much more common to see the definition in the contravariant basis

# Mathematical and physical conventions

> The conventions adopted in gyselalibxx for describing physical quantities such as scalar fields, vector fields and tensors are detailed hereafter. 

Reviewer 2: What is the name of the library? Is it not "Gyselalib++"?

Reviewer 3: I agree, Gyselalib++ is the official name

Author: done 564eea9d4d11cfc40c7d1f5e51fc1351f2c439c7

![Coordinate curves, surfaces and axes in a three-dimensional space](./curvilinear_coordinates_def.png "")

The position of any point in space can be written as

```math
\overrightarrow{r} = x^i e_i,
```

where we introduced the unit vectors of the orthonormal Cartesian basis $`\{e_i\}`$. The position vector $`\overrightarrow{r}`$ is not to be confused with the radial coordinate $`r`$. We rely on the Einstein summation in the above equation and in the following, i.e. we consider that repeated indices indicate a summation. See the [wikipedia page on Einstein summation](https://en.wikipedia.org/wiki/Einstein_notation) for more details.

## Contravariant and covariant bases
Let us define the *contravariant basis* $`\{b_i\}`$ associated to the $`\{q^i\}`$ coordinates by

```math
b_i = \frac{\partial \overrightarrow{r}}{\partial q^i} = \frac{\partial x^j}{\partial q^i} e_j,
```

> and the *covariant basis* $`\{b^i\}`$ by

Reviewer 2: ```suggestion
and its dual basis, the *covariant basis* $`\{b^i\}`$ by
```
?

```math
b^i = \nabla q^i = \frac{\partial q^i}{\partial x^j} e_j.
```

These two bases are *local bases*, in the sense that the $`b_i`$ and $`b^i`$ vectors depend on the considered position in space, i.e. 
we have $`b_i(x^1, \ldots , x^N)`$ and $`b^i(x^1, \ldots , x^N)`$. A geometrical interpretation of these two bases reads as follows. The *covariant* unit vector $`b^i`$ is orthogonal to the coordinate surface $`q^i = \text{constant}`$, while the *contravariant* unit vector $`b_i`$ is locally tangent to the coordinate curve associated with the $`q^i`$ coordinate. This is the situation depicted in the following picture.

> ![Geometrical interpretation of the contravariant and covariant bases vectors](./curvilinear_coordinates_contravariant_covariant_bases.png)

Reviewer 2: The image is not displayed 
![image](https://github.com/user-attachments/assets/05fb5633-7495-4dfe-9e1d-e2d8a7b9c94c)

Author: thanks for spotting the error, it should be fixed now !

Note that neither contravariant nor covariant bases form orthonormal vector sets in general. Additionally note that in the case of Cartesian coordinates covariant and contravariant bases are the same. In general, the following property holds 
```math

> b^i \cdot b_j = \delta_{ij}, 

Reviewer 3: Scalar product to be removed:
```suggestion
b^i b_j = \delta_{ij}, 
```
The scalar product has not yet been defined and if it had it would add an additional layer of complexity here

Reviewer 3: Never mind this comment is not relevant as `b` is a vector

Reviewer 3: Maybe clearer if vectors are bolded

Author: done 32b9436116fb61f4bc47fd82f6fc04d60e0bf6fc

```

> with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`A`$. This vector may be expressed in either contravariant or covariant bases as

Reviewer 2: ```suggestion
with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`A \in \mathbb{R}^N`$. This vector may be expressed in either contravariant or covariant bases as
```
?

> with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`A`$. This vector may be expressed in either contravariant or covariant bases as

Reviewer 2: ```suggestion
with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`A`$. This vector may be expressed in either covariant or contravariant bases as
```
To follow the same order as below. It is easier for me because I never remember which correspond to which one.

Author: done bf9e6d6a2890ae2ed55693671f00cebd87a78900

> with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`A \in \mathbb{R}^N`$. This vector may be expressed in either covariant or contravariant bases as

Reviewer 2: ```suggestion
with $`\cdot`$ the dot product operator and the Kronecker delta $`\delta_{ij} = 1`$ if $`i= j`$, and $`\delta_{ij} = 0`$ otherwise. Let us now consider a vector $`\mathbf{A} \in \mathbb{R}^N`$. This vector may be expressed in either covariant or contravariant bases as
```

```math

> A = A_i \mathbf{b}^i = A^i \mathbf{b}_i, 

Reviewer 2: ```suggestion
\mathbf{A} = A_i \mathbf{b}^i = A^i \mathbf{b}_i, 
```

```

Where 
- $`\{A_i\}`$ are the components of $`A`$ in *the covariant basis*. We call these *the covariant components* of $`A`$;

> - $`\{A^i\}`$ are the components of $`A`$ in *the contravariant basis*. We call these *the contravariant components* of $`A`$;

Reviewer 2: ```suggestion
- $`\{A^i\}`$ are the components of $`\mathbf{A}`$ in *the contravariant basis*. We call these *the contravariant components* of $`A`$;
```

```math
J\left.^{i}_{\;j}\right. = \frac{\partial x^i}{\partial q^j},
```

while the inverse Jacobian $`J^{-1}`$ is defined by its components $`J^{-1}\left.^{i}_{\;j}\right.`$ as 

```math
J^{-1}\left.^{i}_{\;j}\right. = \frac{\partial q^i}{\partial x^j}.
```

Both Jacobian and metric tensor relate to each other as

```math
G = J^{T}J.
```

The Jacobian of a curvilinear coordinate transformation can be used to relate the components of a vector expressed in the Cartesian basis $`\{e_i\}`$ to the components of the vector expressed in the contravariant basis $`\{b_i\}`$ associated with the curvilinear coordinate system. More precisely, let us write a vector $`A`$ as 

```math
A = A^i_\text{c} e_i = A^i b_i, 
```

Where the "c" subscript indicates that the considered components is computed in the Cartesian basis. It can be shown using the chain rule that the following equality holds

```math
A^i_\text{c} = J\left.^{i}_{\;j}\right. A^j. 
```

Note that relating the components of the vector $`A`$ expressed in the covariant basis is less straightforward as this involves the metric tensor, i.e.  

```math

> A^i_\text{c} = J\left.^{i}_{\;j}\right. g^{jk}A_k. 

Reviewer 2: Is it an habit to write the components of the _G_ matrix is lower case and the components of the _J_ matrix in capital letter?

Reviewer 3: I have always seen it written in this way too

Reviewer 2: Ok :+1:

```

### Curvilinear to curvilinear

> Let us now consider the more general case where one seeks to relate two curvilinear coordinate system $`\{q^i\}`$ and $`\{p^i\}`$. We denote by $`J\{q\to p\}`$ the Jacobian of the transformation from the coordinate system $`\{q^i\}`$ to $`\{p^i\}`$. The components of the Jacobian and of its inverse are defined similarly as above as 

Reviewer 2: ```suggestion
Let us now consider the more general case where one seeks to relate two curvilinear coordinate systems $`\{q^i\}`$ and $`\{p^i\}`$. We denote by $`J\{q\to p\}`$ the Jacobian of the transformation from the coordinate system $`\{q^i\}`$ to $`\{p^i\}`$. The components of the Jacobian and of its inverse are defined similarly as above as 
```
?

```math
A^i\{p\} = J\left.^{i}_{\;j}\right.\{q\to p\} A^j\{q\}. 
```

By introducing the metric tensors $`G\{p\}`$  and $`G\{q\}`$ associated with both curvilinear coordinates system, and by writing $`g_{ij}\{p\}`$ and $`g_{ij}\{q\}`$ their elements ($`g^{ij}\{p\}`$ and $`g^{ij}\{q\}`$ the elements of their inverse) one relates the components of $`A`$ expressed in covariant vector basis associated with both curvilinear coordinates system as 

```math
g^{ik}\{p\}A_k\{p\} = J\left.^{i}_{\;j}\right.\{q\to p\}g^{jl}\{p\} A_l\{q\}. 
```

## Differential operators

Hereafter are expressed differential operators in a curvilinear coordinate system $`\{q^i\}`$. 

### Gradient

Let us consider a scalar field $`f`$. The gradient $`\nabla f`$ of such field is defined in the Cartesian basis $`\{e_i\}`$ as

```math
\nabla f = \frac{\partial f}{\partial x^i} e_i. 
```

This quantity can be expressed in the both covariant $`\{b^i\}`$ and contravariant $`\{b_i\}`$  vector basis as

```math
\nabla f = \frac{\partial f}{\partial q^i} b^i = g^{ij} \frac{\partial f}{\partial q^j} b_i. 
```

> Note that the definition that uses the contravariant basis $`\{b_i\}`$ is much more common. 

Reviewer 2: Ah :no_mouth: 
I supposed that we will use the covariant basis for the gradient in the advection equation.
